server/channels/app/custom_profile_attributes.go

```diff
@@ -159,19 +161,34 @@ func (a *App) ListCPAValues(userID string) ([]*model.PropertyValue, *model.AppEr
 		return nil, model.NewAppError("GetCPAFields", "app.custom_profile_attributes.cpa_group_id.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
 	}
 
-	opts := model.PropertyValueSearchOpts{
-		GroupID:        groupID,
-		TargetID:       userID,
-		Page:           0,
-		PerPage:        999999,
-		IncludeDeleted: false,
-	}
-	fields, err := a.Srv().propertyService.SearchPropertyValues(opts)
-	if err != nil {
-		return nil, model.NewAppError("ListCPAValues", "app.custom_profile_attributes.list_property_values.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	var allValues []*model.PropertyValue
+	cursor := model.PropertyValueSearchCursor{}
+	for {
```

Comment:
This entire for-loop is basically just moving the load from the DB to the app. We are appending to the slice until we have all the values. So all that's doing is translate a big SQL query into multiple small ones. It will still run out of memory on the app server if you have sufficiently high number of property values for a user.

The solution is to paginate it from the client side itself. In no case, can we send the entire data from the server to the client. Or if we can't paginate it from the client, we have to stream the value to the client. i.e. - get one page of data, write it to network, get the next page, write that .. and so on.

Reply:
Limiting to 20 entries and dropping the loop sounds good.